app/models/events/offender_event.rb

```diff
@@ -0,0 +1,65 @@
+module Events
+  class OffenderEvent < ApplicationRecord
+    # A custom attribute type to represent the metadata field as its own model object (to support validation)
+    class MetadataType < ActiveRecord::Type::Value
```

Comment:
Try using gem [jsonb_accessor](https://github.com/madeintandem/jsonb_accessor) instead of all this hand-rolled stuff